drivers/api/no_os_i3c.c

```diff
@@ -0,0 +1,263 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include <inttypes.h>
+#include <stdlib.h>
+#include "no_os_i3c.h"
+#include "no_os_error.h"
+#include "no_os_mutex.h"
+#include "no_os_alloc.h"
+
+/**
+ * @brief Initialize the I3C communication peripheral.
+ * @param desc - The I3C descriptor.
+ * @param param - The structure that contains the I3C parameters.
+ * @param param - The device id of the I3C controller to initialize.
+ * @return 0 in case of success, -1 otherwise.
+ */
+int32_t no_os_i3c_init(struct no_os_i3c_master_desc **desc,
+			const struct no_os_i3c_platform_ops *ops,
+			uint8_t device_id)
+{
+	int32_t ret;
+
+	if (!ops)
+		return -EINVAL;
+
+	if (!ops->i3c_ops_init)
+		return -ENOSYS;
+
+	/*
+	 * Execute the platform specific init routine, which
+	 * might configure its internal logic to match the bus limitations.
+	 */
+	ret = ops->i3c_ops_init(desc, ops, device_id);
+	if (ret)
+		goto err_init;
+
+	return 0;
+
+
+err_init:
+	no_os_i3c_remove(*desc);
+
+	return ret;
+}
+
+/**
+ * @brief Attach I3C descriptor to driver pointer.
+ * @param desc - The I3C descriptor.
+ * @param param - The structure that contains the I3C parameters, to match pid.
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t no_os_i3c_init_driver(struct no_os_i3c_slave_desc **desc,
+		       const struct no_os_i3c_init_param *param)
+{
+	struct no_os_i3c_slave_desc **it = i3c_devs_desc;
+	for (; it < i3c_devs_stop_addr; it++) {
+		if (!*it)
+			continue;
+
+		if ((*it)->pid == param->pid)
+			break;
+	}
+	if (it == i3c_devs_stop_addr)
+		return -ENXIO;
+
+	if (!(*it)->is_attached)
+		return -EPERM;
+
+	*desc = *it;
+	return 0;
+}
+
+/**
+ * @brief Do I3C DAA to assign the dynamic addresses.
+ * @param devs- The structure that contains peripherals I3C parameters.
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t no_os_i3c_do_daa(struct no_os_i3c_master_desc *desc)
+{
+	int32_t ret;
+	if (!desc || !desc->platform_ops)
+		return -EINVAL;
+
+	if (!desc->platform_ops->i3c_ops_do_daa)
+		return -ENOSYS;
+
+	/*
+	 * Attach all devices with static address.
+	 */
+	struct no_os_i3c_slave_desc **it = i3c_devs_desc;
```

Comment:
`i3c_devs_desc` is a extern defined in `projects/ad405x/common_data.c`.
Like others variables such as  `i3c_devs_param`, the extern statement is placed in `no_os_i3c.h`.

It was done like this to allow a compile-time defined length and resolution of these variables, depending on the # of devices in the `common_data.c`.